### PR TITLE
Rate limit sessions for throttling control

### PR DESCRIPTION
**Pull request by kostrse**
When using service identity authentication in datasource, query requests originating from different users share the same identity, and may be seen my the resource endpoint as coming from the same user.

I may occasionally trigger throttling problems in large Grafana instances with popular dashboards.

This PR introduces an additional session header `X-RateLimit-Session`, different for each Grafana user, which provide extra information to the resource endpoint regarding the different user sessions which share the same service identity for authentication.

This session header needs to be enabled by the datasource:
```go
authOpts := azhttpclient.NewAuthOptions(azureSettings)
authOpts.Scopes(scopes)

// Enable session header
authOpts.AddRateLimitSession(true)

azhttpclient.AddAzureAuthentication(clientOpts, authOpts, credentials)
```

To make sessions work, datasource should also include context of the signed-in user to the request context, as implemented in the ADX datasource [here](https://github.com/grafana/azure-data-explorer-datasource/blob/afe58c2de6716b61b25c69ebcd981c1530189c97/pkg/azuredx/datasource.go#L78) or [here](https://github.com/grafana/azure-data-explorer-datasource/blob/afe58c2de6716b61b25c69ebcd981c1530189c97/pkg/azuredx/datasource.go#L96). This allows the Azure middleware to use the passed user's information to generate session ID. 

Resource endpoint should know about the header and use it for its rate limit calculations.

### CODE REVIEW

File: azhttpclient/middleware.go [hunk 1]
```diff
@@ -27,31 +28,49 @@ func AzureMiddleware(authOpts *AuthOptions, credentials azcredentials.AzureCrede
...
 			return nil, fmt.Errorf("failed to retrieve Azure access token: %w", err)
 		}
 		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
 
+		if sessionProvider != nil {
+			sessionId, err := sessionProvider.GetSessionId(reqContext)
+			if err != nil {
+				return nil, fmt.Errorf("failed to obtain user session: %w", err)
+			} else if sessionId != "" {
+				req.Header.Set("X-RateLimit-Session", sessionId)
```
Comment (Ohad-Fein) [inline azhttpclient/middleware.go:70]: x-ms-ratelimit-id

File: azhttpclient/middleware.go [hunk 2]
```diff
@@ -27,31 +28,49 @@ func AzureMiddleware(authOpts *AuthOptions, credentials azcredentials.AzureCrede
...
+		token, err := tokenProvider.GetAccessToken(reqContext, scopes)
 		if err != nil {
 			return nil, fmt.Errorf("failed to retrieve Azure access token: %w", err)
 		}
 		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
 
+		if sessionProvider != nil {
+			sessionId, err := sessionProvider.GetSessionId(reqContext)
+			if err != nil {
+				return nil, fmt.Errorf("failed to obtain user session: %w", err)
```
Comment (aangelisc) [inline azhttpclient/middleware.go:68]: I think we should log the error here but not fail the request if there is an error generating the `sessionId` value.

**Review by aangelisc — APPROVED** [review]
One nit in the comments to be addressed before we merge this 😊

File: azhttpclient/middleware.go (hunk 2 above)
Comment (kostrse) [reply azhttpclient/middleware.go:68]: This error is only going to appear due to developer error, improper configuration of datasource http client.
e.g. when the rate limit session logic was enabled in code but user context wasn't configured.

It isn't supposed to transiently appear at runtime.

**Review by aangelisc — APPROVED** [review]

**Pull request merged**